### PR TITLE
Set driver container memory and cpu limits

### PR DESCRIPTION
**Pull request by 88manpreet**
Setting the docker container memory limit to 2x of the spark driver limit.
Cpu limit I am currently setting the same as the spark.driver.cores.
If needs to be higher kindly let me know.
I have added the manual test results in the ticket COREML-2743
I will test one of the heavy batches in production once the code is reviewed.

### CODE REVIEW

File: paasta_tools/cli/cmds/spark_run.py [hunk 1]
```diff
@@ -606,6 +636,30 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+        # For any reason it fails, continue with default value
+        log.error(
+            f"Failed to parse docker memory limit. Error: {e}. Example values: 1g, 200m."
+            " Setting a default value {DEFAULT_MAX_DOCKER_MEMORY_LIMIT}"
+        )
+        docker_memory_limit = DEFAULT_MAX_DOCKER_MEMORY_LIMIT
+
+    docker_cpu_limit = str(
+        int(spark_conf.get("spark.driver.cores", DEFAULT_DRIVER_CORES_BY_SPARK))
+    )
```
Comment (jfongatyelp) [inline paasta_tools/cli/cmds/spark_run.py:659]: Do we ensure spark.driver.cores is convertible to an int anywhere before this, or should we catch exceptions here too like we do for memory?

File: paasta_tools/cli/cmds/spark_run.py [hunk 2]
```diff
@@ -894,6 +954,16 @@ def paasta_spark_run(args):
...
+
+    # Get driver-memory and cores
+    sub_cmds = args.cmd.split(" ")  # spark.driver.memory=10g
+    for cmd in sub_cmds:
+        if cmd.startswith("spark.driver.memory") or cmd.startswith(
+            "spark.driver.cores"
+        ):
+            key, value = cmd.split("=")
+            user_spark_opts[key] = value
+
```
Comment (jfongatyelp) [inline paasta_tools/cli/cmds/spark_run.py:966]: Is this config not always part of `--spark args '...'` in the cmd? If it is, could we not grab this out of the spark_args that we parsed just above this? It seems like [_parse_user_spark_args](https://github.com/Yelp/paasta/blob/012a17ea344e6c932a05607af105a03fdf2ed71c/paasta_tools/cli/cmds/spark_run.py#L545-L555) would just parse anything out of there and throw it into `user_spark_opts`?

**Review by jfongatyelp — APPROVED** [review]

File: paasta_tools/cli/cmds/spark_run.py [hunk 3]
```diff
@@ -56,6 +56,12 @@
 CLUSTER_MANAGER_MESOS = "mesos"
 CLUSTER_MANAGER_K8S = "kubernetes"
 CLUSTER_MANAGERS = {CLUSTER_MANAGER_MESOS, CLUSTER_MANAGER_K8S}
+# Reference: https://spark.apache.org/docs/latest/configuration.html#application-properties
+DEFAULT_DRIVER_CORES_BY_SPARK = 1
+DEFAULT_DRIVER_MEMORY_BY_SPARK = "1g"
```
Comment (gcollura) [inline paasta_tools/cli/cmds/spark_run.py:61]: I'm actually worried in making this the default at the moment. There are a lot of batches that don't specify `spark.driver.cores` and/or `spark.driver.memory` that might fail if this change is pushed. What is the pushplan for this change?

File: paasta_tools/cli/cmds/spark_run.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gcollura) [reply paasta_tools/cli/cmds/spark_run.py:966]: Looks like @jfongatyelp beat me to this comment :)

File: paasta_tools/cli/cmds/spark_run.py (hunk 3 above)
Comment (chelspea) [reply paasta_tools/cli/cmds/spark_run.py:61]: Do have the same concern as above

Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:61]: These default values are the one provided by spark for the above configs. 
The behavior of the batch should be same irrespective of these configs not provided or provided with above values. 
Manual test results in the ticket.

tl;dr when batch doesn't provide these configs:
* Docker container memory limit as 2g
* Docker cpu limit same 1

My only concern is, if should have the core limit > the one set by spark driver.

File: paasta_tools/cli/cmds/spark_run.py (hunk 1 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:659]: 1. If spark.driver.cores is convertible to int before this, it would still work, right?
2. I tested in manual test cases, and didn't see any issues. Ticket comments for results
3. I thought about catching the error here too, but this seems like to basic operation. Only cases I see it failing is spark.driver.cores is provided as (non-int) string or float (or a different type than int-str). 
Happy to catch it for extra caution.

File: paasta_tools/cli/cmds/spark_run.py (hunk 2 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:966]: This was added to cover all edge cases.
If users doesn't provide driver.cores as part of `--spark-args` but provide as part of `spark-submit` command, it would fail.
Example: `spark-submit --conf spark.driver.cores=12g` would set container memory as `2g`. 
I caught this in detailed manual runs, details in the ticket comment.

Happy to revisit, if I still missed something :)

[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:966]: Check the response above :)

File: paasta_tools/cli/cmds/spark_run.py (hunk 1 above)
Comment (jfongatyelp) [reply paasta_tools/cli/cmds/spark_run.py:659]: One more note about this is that `docker run` will flat out fail to run if we specify more `--cpus` than the host has total -- with our batch boxes having only 8vcpu, we'd want to be a little careful about this; some configs in yelpsoa-configs already appear to specify spark.driver.cores=8, thankfully none more than 8.

Is there already documentation to tell folks not to use >8? If not, the user experience might be slightly unfriendly, where they only find out their tronjob's config is not valid when the job fails.

File: paasta_tools/cli/cmds/spark_run.py [hunk 4]
```diff
@@ -56,6 +56,12 @@
...
+# Extra room for memory overhead and for any other running inside containe
```
Comment (nemacysts) [inline paasta_tools/cli/cmds/spark_run.py:62]: super-nit
```suggestion
# Extra room for memory overhead and for any other running inside container
```

File: paasta_tools/cli/cmds/spark_run.py [hunk 5]
```diff
@@ -606,6 +636,30 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
     return spark_app_name
 
 
+def _calculate_docker_memory_cpu_limit(spark_conf):
```
Comment (nemacysts) [inline paasta_tools/cli/cmds/spark_run.py:639]: nit: could we add types to any new functions we add? imo, it'd also be nice to separate this into two functions so that you don't need to look at the code to know what order memory/cpu are in

File: paasta_tools/cli/cmds/spark_run.py [hunk 6]
```diff
@@ -606,6 +636,30 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+    try:
+
+        docker_memory_limit_str = spark_conf.get(
+            "spark.driver.memory", DEFAULT_DRIVER_MEMORY_BY_SPARK
+        )
+        match = re.match(r"([0-9]+)([a-z]*)", docker_memory_limit_str)
```
Comment (nemacysts) [inline paasta_tools/cli/cmds/spark_run.py:645]: should the memory unit party of the regex be the subset of allowed units? e.g., 
```Python
        match = re.match(r"([0-9]+)(mb|gb)?", docker_memory_limit_str.lower())
```

or do we have a pre-commit hook or something that ensures that people don't make mistakes here?

File: paasta_tools/cli/cmds/spark_run.py [hunk 7]
```diff
@@ -56,6 +56,12 @@
 CLUSTER_MANAGER_MESOS = "mesos"
 CLUSTER_MANAGER_K8S = "kubernetes"
 CLUSTER_MANAGERS = {CLUSTER_MANAGER_MESOS, CLUSTER_MANAGER_K8S}
+# Reference: https://spark.apache.org/docs/latest/configuration.html#application-properties
+DEFAULT_DRIVER_CORES_BY_SPARK = 1
+DEFAULT_DRIVER_MEMORY_BY_SPARK = "1g"
+# Extra room for memory overhead and for any other running inside containe
+DOCKER_RESOURCE_ADJUSTMENT_FACTOR = 2
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:63]: There is driver.memory.overheadfactor which also uses driver memory (default 10%) and some users can have higher value. Accounting that into the calculation can make the code complicated.

Question: If the docker container memory limit is set to 4G (let's say), and the box has 40G storage. Does that mean only 10 docker containers will be allocated at max? Or 20 docker containers could be allocated too, but none of them would exceed 4G.

File: paasta_tools/cli/cmds/spark_run.py (hunk 6 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:645]: If the units are wrong, the spark driver will fail and/or so does creating the container. 
I think adding this check will be an overkill here, and would make it harder to debug in future and needs to consider all possible edge cases (e.g. 1000 is valid value too just like 1000m).

File: paasta_tools/cli/cmds/spark_run.py (hunk 2 above)
Comment (jfongatyelp) [reply paasta_tools/cli/cmds/spark_run.py:966]: Ah, I didn't realize people could specify this in the spark-submit command instead, thanks for the explanation.

Comment (jfongatyelp) [reply paasta_tools/cli/cmds/spark_run.py:966]: Could you add a comment indicating this for future us?

File: paasta_tools/cli/cmds/spark_run.py (hunk 1 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:659]: Good Catch. I tested this max value case for driver-memory (in the ticket) where in `docker run` just set the value to max memory of the box by default. Strange to me, that it didn't do the same for `cpus' too.
I will test the same for the max value too.

What should be the max value you suggest? This is considering that many instance types are being used for batch and stable_batch boxes?
Is 8vcpu's standard and reliable long-term consistent value to be used to check for max allowed value?

Comment (jfongatyelp) [reply paasta_tools/cli/cmds/spark_run.py:659]: 8 seems fine to leave as the max although it's literally the maximum available on these hosts. The batch and stable_batch boxes aren't involved with the driver, and the boxes that run our drivers are all only 1 instance type right now w/ only 8vcpu. 

So since we're counting on drivers moving to k8s soon (where we can actually schedule based on resources requested), it might be sufficient just to leave this and update documentation, maybe add a pre-commit to yelpsoa-configs.

Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:659]: Ok. I will leave it like that. 
Interesting in local test, when i tried setting 10 cores. It just let me and continued running.
`docker inspect 6c5e647b4632 | grep Nano
            "NanoCpus": 10000000000,`
Maybe it becomes more of the useless value when greater than max-cores as opposed to an issue. 
Which I think might be ok.

File: paasta_tools/cli/cmds/spark_run.py (hunk 2 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:966]: Added.

File: paasta_tools/cli/cmds/spark_run.py (hunk 3 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:61]: @gcollura , @chelspea let me know if I can resolve this comment?

File: paasta_tools/cli/cmds/spark_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:63]: I make it configurable optionally.

[the remark this answers is not in the document]
Comment (nemacysts) [reply paasta_tools/cli/cmds/spark_run.py:63]: for your question: since we're running these containers directly, it'd be the latter - there's nothing stopping Tron's SSH code from starting more docker containers (with a 4GB memory limit) than would actually fit onto a box. 

(e.g., it would happily try to start 100 containers with a memory limit of 4GB)

Setting this limit won't solve all the problems with running the drivers on bare-metal because of this, but it should significantly improve things when one driver starts hogging all the memory

File: paasta_tools/cli/cmds/spark_run.py [hunk 8]
```diff
@@ -894,6 +992,18 @@ def paasta_spark_run(args):
     user_spark_opts = _parse_user_spark_args(
         args.spark_args, pod_template_path, args.enable_compact_bin_packing
     )
+
+    # Get driver-memory and cores
+    sub_cmds = args.cmd.split(" ")  # spark.driver.memory=10g
```
[the remark this answers is not in the document]
Comment (nemacysts) [reply paasta_tools/cli/cmds/spark_run.py:997]: this also seems a little brittle, but that's a problem for future-us I guess (something like providing native soaconfigs support for this would be nice)

File: paasta_tools/cli/cmds/spark_run.py [hunk 9]
```diff
@@ -606,6 +650,53 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+    2. --spark-args or spark-submit: spark.driver.cores
+    3. Default
+    """
+    if cpu_limit:
+        docker_cpu_limit = cpu_limit
+    else:
+        docker_cpu_limit = str(
+            int(spark_conf.get("spark.driver.cores", DEFAULT_DRIVER_CORES_BY_SPARK))
+        )
+    return docker_cpu_limit
```
Comment (nemacysts) [inline paasta_tools/cli/cmds/spark_run.py:697]: feel free to drop this suggestion:
```suggestion
    return cpu_limit if cpu_limit else spark_conf.get("spark.driver.cores", str(DEFAULT_DRIVER_CORES_BY_SPARK))
```

(from my understanding `spark.driver.cores` will always be a string, so we really just want to cast the default core value)

File: paasta_tools/cli/cmds/spark_run.py (hunk 5 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py]: Added

File: paasta_tools/cli/cmds/spark_run.py (hunk 1 above)
Comment (jfongatyelp) [reply paasta_tools/cli/cmds/spark_run.py]: ^This test was done on a host that had >10 cpu available; can you try again either specifying more cpu than your test host has or testing on one of the batch boxes?

my concern here is that for most `spark-run` tron jobs, the driver will be run on a host that has only 8 cpu available, so we either a) give users a bad experience where they can set up yelpsoa-configs with >8 spark.driver.cores for their job, then only find out it fails on the first scheduled run, or b) add some sort of safety here to just set this value to maximum for the host it's running on

File: paasta_tools/cli/cmds/spark_run.py [hunk 10]
```diff
@@ -606,6 +650,53 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+) -> str:
+    """ In Order of preference:
+    1. Argument: --docker-memory-limit
+    2. --spark-args or spark-submit: spark.driver.memory
+    3. Default
+    """
+    try:
+        if memory_limit:
+            docker_memory_limit_str = memory_limit
+            adjustment_factor = 1
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:664]: The reason I kept it is the same you mentioned later.
Making sure we apply the same validation rules to user inputs through command-line args or configs. 
I believe docker will validate anyways. I can simplify this.

File: paasta_tools/cli/cmds/spark_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:63]: Makes sense. That's what I thought too. Thanks for confirming :)

File: paasta_tools/cli/cmds/spark_run.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:991]: I had added the detailed comment at line #993-994. I will move it upwards. Hopefully, that is more clearer

[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:991]: > just double-checking: there's no values for any of these config options that could potentially have a space in them?
This is correct. In fact I could remove this too. There is no real use-case at Yelp being used in this.
I added it to take care of future possible edge cases.

File: tests/cli/test_cmds_spark_run.py [hunk 11]
```diff
@@ -42,12 +44,21 @@ def test_get_docker_run_cmd(mock_getegid, mock_geteuid):
...
+        volumes,
+        env,
+        docker_img,
+        docker_cmd,
+        nvidia,
+        docker_memory_limit,
+        docker_cpu_limit,
     )
 
-    assert actual[5:] == [
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply tests/cli/test_cmds_spark_run.py:50]: Seems like an overkill check to me, as we are setting it for all possible edge cases.
This should get caught in unit-tests in case of any failure.

File: paasta_tools/cli/cmds/spark_run.py (hunk 1 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py]: I tried with > 10 cpus. It failed. I added more details in the help section for better user experience. This has been discussed offline with @jfongatyelp

File: paasta_tools/cli/cmds/spark_run.py (hunk 4 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:62]: Fixed

File: paasta_tools/cli/cmds/spark_run.py (hunk 9 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:700]: Added back.

File: paasta_tools/cli/cmds/spark_run.py [hunk 12]
```diff
@@ -143,6 +149,20 @@ def add_subparser(subparsers):
...
+            "Set docker memory limit. Should be greater than driver memory. Defaults to 2x driver memory. Example: 2g, 500m, Max: 64g"
+        ),
+        default=False,
+    )
+    list_parser.add_argument(
+        "--docker-cpu-limit",
+        help=(
+            "Set docker cpus limit. Should be greater than driver cores. Defaults to 1x driver cores."
+        ),
+        default=False,
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py]: Added back.

File: paasta_tools/cli/cmds/spark_run.py [hunk 13]
```diff
@@ -606,6 +650,53 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+        log.error(
+            f"Failed to parse docker memory limit. Error: {e}. Example values: 1g, 200m."
+            " Setting a default value {DEFAULT_MAX_DOCKER_MEMORY_LIMIT}"
+        )
+        docker_memory_limit = DEFAULT_MAX_DOCKER_MEMORY_LIMIT
+
+    return docker_memory_limit
+
+
+def _calculate_docker_cpu_limit(spark_conf: Mapping[str, str], cpu_limit: str) -> str:
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py]: Added back.

File: paasta_tools/cli/cmds/spark_run.py [hunk 14]
```diff
@@ -606,6 +650,53 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
     return spark_app_name
 
 
+def _calculate_docker_memory_limit(
+    spark_conf: Mapping[str, str], memory_limit: str
```
[the remark this answers is not in the document]
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py]: Added back.

File: paasta_tools/cli/cmds/spark_run.py [hunk 15]
```diff
@@ -606,6 +652,52 @@ def get_spark_app_name(original_docker_cmd: Union[Any, str, List[str]]) -> str:
...
+        memory_val = int(match[1]) * adjustment_factor
+        memory_unit = match[2]
+        docker_memory_limit = f"{memory_val}{memory_unit}"
+    except Exception as e:
+        # For any reason it fails, continue with default value
+        print(
+            f"Failed to parse docker memory limit. Error: {e}. Example values: 1g, 200m."
+            " Setting a default value {DEFAULT_MAX_DOCKER_MEMORY_LIMIT}"
+        )
+        docker_memory_limit = DEFAULT_MAX_DOCKER_MEMORY_LIMIT
```
Comment (nemacysts) [inline paasta_tools/cli/cmds/spark_run.py:681]: thinking about this more: should we just fail here? we're already going to fail if people do something weird with the cpu limit too

File: paasta_tools/cli/cmds/spark_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nemacysts) [reply paasta_tools/cli/cmds/spark_run.py:63]: I was thinking of making this configurable through Puppet (using https://github.com/Yelp/paasta/blob/master/paasta_tools/utils.py#L1847) so that we don't need to release new paasta versions to tweak this factor - but this is fine for now

**Comment by gcollura** [comment]
I don't have much to add, @nemacysts left great comments.
One thought though, Tron on Paasta has `cpus/mem` options, should we add a feature in Tron where cpu and mem for the driver can be set through tron config as well?

File: paasta_tools/cli/cmds/spark_run.py (hunk 15 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:681]: Most users specify high memory limit as part of over-provisioning for successful run of the job.
Having it failed IMHO is not a smooth user-experience for both onpoint and end-user (delay, notice, reach onpoint, change memory limits, calculate the right value, knowing the max value).

Even if it fails, the user would need to specify the max driver limit of the container (or half of it). 
Having it as a warning is good enough to be aware of and take decisions if required. 
I would rather prefer to keep it like it this.

**Comment by 88manpreet** [comment]
> I don't have much to add, @nemacysts left great comments. One thought though, Tron on Paasta has `cpus/mem` options, should we add a feature in Tron where cpu and mem for the driver can be set through tron config as well?

I think if we can find the strong case for generic usage, it would be useful to have.
I think we would want to make our user experience less as less confusing with multiple parameters as possible.
Right now, they can do it with `--docker-<cpu|memory>-limit` (which is still configurable). But I doubt it is going to required much. The default values should suffice for 90% of the cases, I think. 

Still something good to keep in our mind, for future versions.

File: paasta_tools/cli/cmds/spark_run.py (hunk 15 above)
Comment (nemacysts) [reply paasta_tools/cli/cmds/spark_run.py:681]: this block would only get hit if users made a mistake setting up their config tho, right? e.g., if they set `10h` instead of `10g`, we/they would never know that their config is invalid and this would basically mean that we fall back to the status quo where a single spark job can take down a box

**Comment by nemacysts** [comment]
> I don't have much to add, @nemacysts left great comments. One thought though, Tron on Paasta has `cpus/mem` options, should we add a feature in Tron where cpu and mem for the driver can be set through tron config as well?

@gcollura i was actually thinking of adding support for something like this when I start working on the driver-in-k8s stuff for y'all :)

File: paasta_tools/cli/cmds/spark_run.py (hunk 15 above)
Comment (88manpreet) [reply paasta_tools/cli/cmds/spark_run.py:681]: Sorry I misinterpreted your comment in lines with failing when user specifies > than the max allowed limit of the box.
Yes, you are right.
My original reasoning to not fail was, in the goal of not impacting batches doesn't fall into this regex and keep the job running as status-quo. As in WARN, Move-on, fix based on warning for next runs.

I guess I will change this to fail-hard.

Comment (nemacysts) [reply paasta_tools/cli/cmds/spark_run.py]: would it be worth writing a quick script to check all the existing configs to see if there are any existing batches that have incorrect config?

**Review by nemacysts — APPROVED** [review]

**Review by gcollura — APPROVED** [review]

**Review by jfongatyelp — APPROVED** [review]
